editor/editor_command_palette.cpp

```diff
@@ -0,0 +1,300 @@
+		add_command(command_name, *p_key, callable_mp(EditorNode::get_singleton()->get_viewport(), &Viewport::unhandled_input), varray(ev, false), shortcut_text);
+		p_key = unregistered_shortcuts.next(p_key);
+	}
+	unregistered_shortcuts.clear();
+}
+
+Ref<Shortcut> EditorCommandPalette::add_shortcut_command(const String &p_command, const String &p_key, Ref<Shortcut> p_shortcut) {
+	if (is_inside_tree()) {
+		Ref<InputEventShortcut> ev;
+		ev.instantiate();
+		ev->set_shortcut(p_shortcut);
+		String shortcut_text = String(p_shortcut->get_as_text());
+		add_command(p_command, p_key, callable_mp(EditorNode::get_singleton()->get_viewport(), &Viewport::unhandled_input), varray(ev, false), shortcut_text);
+	} else {
+		const String key_name = String(p_key);
+		const String command_name = String(p_command);
+		Pair p_pair = Pair(command_name, p_shortcut);
+		unregistered_shortcuts[key_name] = p_pair;
+	}
+	return p_shortcut;
+}
+
+void EditorCommandPalette::_theme_changed() {
+	command_search_box->set_right_icon(search_options->get_theme_icon("Search", "EditorIcons"));
+}
+
+EditorCommandPalette *EditorCommandPalette::get_singleton() {
+	if (singleton == nullptr) {
+		singleton = memnew(EditorCommandPalette);
+	}
+	return singleton;
+}
+
+EditorCommandPalette::EditorCommandPalette() {
+	VBoxContainer *vbc = memnew(VBoxContainer);
+	vbc->connect("theme_changed", callable_mp(this, &EditorCommandPalette::_theme_changed));
+	add_child(vbc);
+
+	command_search_box = memnew(LineEdit);
+	command_search_box->set_placeholder("search for a command");
+	command_search_box->set_placeholder_alpha(0.5);
+	command_search_box->connect("gui_input", callable_mp(this, &EditorCommandPalette::_sbox_input));
+	command_search_box->connect("text_changed", callable_mp(this, &EditorCommandPalette::_update_command_search));
+	command_search_box->connect("text_submitted", callable_mp(this, &EditorCommandPalette::_confirmed).unbind(1));
+	command_search_box->set_v_size_flags(Control::SIZE_EXPAND_FILL);
+	MarginContainer *margin_container_csb = memnew(MarginContainer);
+	margin_container_csb->add_child(command_search_box);
+	vbc->add_child(margin_container_csb);
+	register_text_enter(command_search_box);
+
+	search_options = memnew(Tree);
+	search_options->connect("item_activated", callable_mp(this, &EditorCommandPalette::_confirmed));
+	search_options->create_item();
+	search_options->set_hide_root(true);
+	search_options->set_hide_folding(true);
+	search_options->add_theme_constant_override("draw_guides", 1);
+	search_options->set_columns(2);
+	search_options->set_v_size_flags(Control::SIZE_EXPAND_FILL);
+	search_options->set_h_size_flags(Control::SIZE_EXPAND_FILL);
+	search_options->set_column_custom_minimum_width(0, int(8 * EDSCALE));
+
+	vbc->add_child(search_options, true);
+
+	set_hide_on_ok(false);
+}
+
+Ref<Shortcut> ED_SHORTCUT_AND_COMMAND(const String &p_path, const String &p_name, uint32_t p_keycode, String p_command_name) {
```

Comment:
I forgot to mention, I don't really like this shortcut AND command function. I don't think it should have the side effect of also registering the command in editor settings. Personally I think they should be separate. More verbose, yes, but the function shouldn't have external side effects IMO.

```cpp
p->add_shortcut(ED_SHORTCUT_AND_COMMAND("editor/quick_open_script", TTR("Quick Open Script..."), KEY_MASK_CMD + KEY_MASK_ALT + KEY_O), FILE_QUICK_OPEN_SCRIPT);

// becomes...

Ref<Shortcut> sc = ED_SHORTCUT("editor/quick_open_script", TTR("Quick Open Script..."), KEY_MASK_CMD + KEY_MASK_ALT + KEY_O)
EditorCommandPalette::get_singleton()->add_shortcut_command(sc->get_name(), "editor/quick_open_script", p_shortcut);
p->add_shortcut(sc, FILE_QUICK_OPEN_SCRIPT);


```

Reply:
> I forgot to mention, I don't really like this shortcut AND command function. I don't think it should have the side effect of also registering the command in editor settings. Personally I think they should be separate. More verbose, yes, but the function shouldn't have external side effects IMO.

I disagree on that. This is a function that is supposed to be called very often in the codebase, going for 2 functions would be a pain to deal with, as it would mean 3 lines of codes everytime you need both a shortcut and a command. The most common use case is to add both, so I think a dedicated function here makes sense.

Reply:
IMO if the most common use case is to add both I would have just made `ED_SHORTCUT` do that, and added `ED_SHORTCUT_NO_COMMAND` for the few shortcuts which should not be exposed as commands.

Reply:
Well, I kind of agree, but I meant the most common use case compared to adding only the command (without registering the shortcut). For now I don't know which proportion of the shortcuts makes sense to be added commands too.

Reply:
Yeah, I see where you are coming from. Personally I try to avoid functions which have 'and' in the name (ideally following srp) or make calls to external classes.

At the end of the day this is fine (and very nitpicky on my end) because the use of this method is limited to a very specific case.

Reply:
@EricEzaM To be fair, in general I would agree with you. But I think here the case is special because we kind of use this function as kind of a macro to avoid duplication of code. In the long run I believe it avoids mistakes.
Also, right now the macro is in `editor_command_palette.h`,  but I agree it could also be move to another place to avoid including the editor shortcuts header there. But well, not sure it would have been worth to create a file only for that.